h2o-core/src/main/java/hex/grid/HyperSpaceWalker.java

```diff
@@ -437,6 +440,123 @@ public void modelFailed(Model failedModel) {
     }
   } // class CartesianWalker
 
+
+  /**
+   * Hyperparameter space walker which visits each combination of hyperparameters in order.
+   */
+  public static class MaterializedRandomWalker<MP extends Model.Parameters>
+          extends BaseWalker<MP, HyperSpaceSearchCriteria.RandomDiscreteValueSearchCriteria> {
+
+    public MaterializedRandomWalker(MP params,
```

Comment:
Is this meant to be accessible to users? The search space for the default AutoML GBM grid search is about 80k. That is a lot of hashmap objects.

Reply:
@michalkurka yes. I'm conscious about the tradeoff between lazy and eager evaluations. That is why it put word `materialised` in the name of the class `MaterializedRandomWalker`. Maybe this walker will not be perfect for all use cases(grids) as any solution in general might not be, but we will definitely need something like this for Bayesian approaches. Later we can change representation to make it more memory efficient. I would say that we can store grid items as just ordered(based on array of grid keys) arrays Object[]. 

Also as it is a Stream we can do lazy materialisation as well. But shuffling will be a difficulty here. We can materialise in batches and shuffle them but it might not be as good shuffling as given whole array of grid items.

Reply:
> Later we can change representation to make it more memory efficient. I would say that we can store grid items as just ordered(based on array of grid keys) arrays Object[].

Why not thinking only in terms of indices?
Param names are already sorted according to `_hyperParamNames`.
You know how many potential values for each `hyperParam`.
So , with simple arithmetic, you can build a unique match between each param combination and an index `i`.
You build an array [0....max], and this is easy to randomize.
Then for each number, you reverse the arithmetic to recover the params name/value, apply the filter and so on...

Reply:
@seb-h2o as I mentioned above I will think about optimisation part once I prepare everything else . It is always a tradeoff..... once you store as indices you will spend more time for applying functions and the more functions you have the more times you will need to convert. We can change representation after we filtered everything - that is what I was thinking about. Let me finish and then we can discuss.... there is a WIP label.

Reply:
It's arithmetic + random access! It's always fast!